File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def lowerbound1(self,nums: List[int], target: int):
        low,high,ans=0,len(nums)-1,-1
        while low<=high:
            mid= (low+high)//2
            if nums[mid]==target:
                ans=mid
            if nums[mid]>=target:
                high=mid-1
            else:
                low=mid+1
        return ans
    def upperbound1(self,nums: List[int], target: int):
        low,high,ans=0,len(nums)-1,-1
        while low<=high:
            mid= (low+high)//2
            if nums[mid]==target:
                ans=mid
            if nums[mid]<=target:
                low=mid+1
            else:
                high=mid-1;
        return ans
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        return (self.lowerbound1(nums,target),self.upperbound1(nums,target))

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        // Here we are given sorted repeated array and we need to find th first and last position of occurance of that element

        int low=0,high=nums.size()-1;
        int ans1=-1,ans2=-1;
        while(low<=high){
            int mid = (low+high)/2;

            if(nums[mid]==target){
                int i=mid,j=mid;
                while(i>=0 && nums[i]==target){
                    ans1 = i;
                    i--;
                }
                while(j<nums.size() && nums[j]==target){
                    ans2=j;
                    j++;
                }
                return {ans1,ans2};
            }

            if(nums[mid]<target){
                low=mid+1;
            }
            else{
                high=mid-1;
            }
        }
        return {ans1,ans2};
    }
};